iron/gpg-packet.c

```diff
@@ -0,0 +1,1150 @@
+    sshbuf_put_u8(body, GPG_SKALGO_AES256);
+    sshbuf_put_u8(body, GPG_SKALGO_AES192);
+    sshbuf_put_u8(body, GPG_SKALGO_AES128);
+    sshbuf_put_u8(body, 0);         //  No fourth preferred SK algorithm
+    sshbuf_put_u8(body, 2);         //  Length of key flags subpacket
+    sshbuf_put_u8(body, GPG_SIG_SUBPKT_KEY_FLAGS);
+    if (pubkey_tag == GPG_TAG_PUBLIC_KEY) {
+        sshbuf_put_u8(body, 0x03);      // Sign + certify
+    } else if (pubkey_tag == GPG_TAG_PUBLIC_SUBKEY) {
+        sshbuf_put_u8(body, 0x0c);      // encrypt
+    } else {
+        sshbuf_put_u8(body, 0x00);
+    }
+    sshbuf_put_u8(body, 2);         //  Length of features subpacket
+    sshbuf_put_u8(body, GPG_SIG_SUBPKT_FEATURES);
+    sshbuf_put_u8(body, 0x01);          // enabled MDC - integrity protection for encrypted data packets
+}
+
+/**
+ *  Adds the little "trailer" sequence into signature hash.
+ *
+ *  Assumes that a hash has been computed on the data being signed. GPG adds this little 6 byte trailer into
+ *  the hash before finalizing it. The length here is the number of bytes in the signature packet from the
+ *  start of the packet (not including the tag/length header) through the end of the hashed data subpackets.
+ *
+ *  @param ctx SHA256 hash to update
+ *  @param len number of bytes in start of signature packet, through the hashed data subpackets
+ */
+static void
+add_signature_trailer(SHA256_CTX * ctx, int len)
+{
+	u_char trailer[6];
+	trailer[0] = GPG_SIG_VERSION;
+	trailer[1] = 0xff;
+	iron_int_to_buf(len, trailer + 2);
+	SHA256_Update(ctx, trailer, 6);
+}
+
+/**
+ *  Compute the hash that goes in the signature packet
+ *
+ *  The hash is over the pubkey packet, uid packet, and the start of the signature packet.
+ *
+ *  @param pubkey_pkt Already generated public key (or subkey) packet
+ *  @param uid_pkt Already generated UID packet, or NULL if no UID packet (signing subkey, for example)
+ *  @param sig_pkt First portion of the signature packet (everything that needs to be hashed)
+ *  @param hash Place to write the hash (at least SHA256_DIGEST_LENGTH bytes)
+ */
+static void
+compute_signature_hash(const gpg_packet * pubkey_pkt, const gpg_packet * uid_pkt, const gpg_packet * sig_pkt,
+                       u_char * hash)
+{
+    SHA256_CTX  ctx;
+    SHA256_Init(&ctx);
+
+    u_char buf[6];
+
+    /* The hash is computed over the entire public key packet, the entire UID packet (except the stupid length is
+     * expanded to four bytes), and the first part of the signature packet. It ends up with a trailer that is the
+     * version, 0xff, and the four-byte length of the hashed data, MSB first.
+     */
+
+    /* First, generate the tag/length for the public key packet, hash that, then hash the packet contents. */
+    int len = generate_gpg_tag_and_size(pubkey_pkt->tag, pubkey_pkt->len, buf);
+    SHA256_Update(&ctx, buf, len);
+    SHA256_Update(&ctx, sshbuf_ptr(pubkey_pkt->data), sshbuf_len(pubkey_pkt->data));
+
```

Comment:
Does the second SHA256_Update not overwrite the &ctx? Does it append results? So this is the sha of the length concatenated with the sha of the data?